pyomo/core/base/block.py

```diff
@@ -1836,11 +1839,31 @@ def __new__(cls, *args, **kwds):
     def __init__(self, *args, **kwargs):
         """Constructor"""
         self._suppress_ctypes = set()
-        self._rule = kwargs.pop('rule', None)
-        self._options = kwargs.pop('options', None)
+        _rule = kwargs.pop('rule', None)
+        _options = kwargs.pop('options', None)
+        # As concrete applies to the Block at declaration time, we will
+        # not use an initializer.
         _concrete = kwargs.pop('concrete', False)
+        # As dense applies to the whole container, we will not use an
+        # initializer
+        self._dense = kwargs.pop('dense', True)
```

Comment:
I don't think I understand the implications of this. Were blocks not dense before?

Reply:
Blocks were dense if you specified a rule (and indexed them with a finite set).  This was unfortunate, as `Complementary` relied on the face that if the rule was omitted the Block was sparse.  This does change the default behavior so that Blocks (indexed by finite sets) are not dense by default.  I don't think that it really changes anything for people, especially as I don't know of anyone who was relying on that particular edge case.